### PR TITLE
feat: place comment

### PR DESCRIPTION
**Pull request by michaldomagalski**
Related to: https://github.com/copotrzebne-pl/copotrzebne-api/issues/43

### CODE REVIEW

File: migrations/20220324080020-add-place-comment.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            allowNull: false,
+          },
+          message: {
+            type: Sequelize.STRING,
+            allowNull: false,
+          },
+          links: {
+            type: Sequelize.ARRAY(Sequelize.TEXT),
+            allowNull: true,
+          },
```
Comment (adrkakol) [inline migrations/20220324080020-add-place-comment.js:28]: Can we set it to empty array by default? Or is Sequelize handling that (when null then returns [])?

File: migrations/20220324080020-add-place-comment.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          place_id: {
+            type: Sequelize.UUID,
+            allowNull: true,
+            references: {
+              model: 'places',
+              key: 'id',
+            },
+            onUpdate: 'set null',
```
Comment (adrkakol) [inline migrations/20220324080020-add-place-comment.js:36]: is this correct? Do we want to have orphaned comments?

File: migrations/20220324080020-add-place-comment.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            onDelete: 'set null',
```
Comment (adrkakol) [inline migrations/20220324080020-add-place-comment.js:37]: Maybe delete cascade?

File: migrations/20220324080020-add-place-comment.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            type: Sequelize.STRING,
+            allowNull: false,
+          },
+          links: {
+            type: Sequelize.ARRAY(Sequelize.TEXT),
+            allowNull: true,
+          },
+          place_id: {
+            type: Sequelize.UUID,
+            allowNull: true,
```
Comment (adrkakol) [inline migrations/20220324080020-add-place-comment.js:31]: I think we don't want orphaned comments, they belong to the place always. If place is removed, we want to remove comment

File: src/comments/comments.controller.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      return comment;
+    });
+  }
+
+  @ApiResponse({ type: Comment, description: 'creates comment and returns created entity' })
+  @SetMetadata(MetadataKey.ALLOWED_ROLES, [UserRole.ADMIN, UserRole.PLACE_MANAGER])
+  @UseGuards(AuthGuard)
+  @Post('/')
+  public async createComment(@Body() commentDto: CreateCommentDto): Promise<Comment | void> {
+    return await this.sequelize.transaction(async (transaction) => {
```
Comment (adrkakol) [inline src/comments/comments.controller.ts:55]: You have to check if !admin user has access to the place to which he creates a comment. Just use places service method `isPlaceManageableByUser`

File: src/comments/comments.controller.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      return comment;
+    });
+  }
+
+  @ApiResponse({ type: Comment, description: 'updates comment and returns updated entity' })
+  @SetMetadata(MetadataKey.ALLOWED_ROLES, [UserRole.ADMIN, UserRole.PLACE_MANAGER])
+  @UseGuards(AuthGuard)
+  @Patch('/:id')
+  public async updateComment(@Param('id') id: string, @Body() commentDto: UpdateCommentDto): Promise<Comment | void> {
+    return await this.sequelize.transaction(async (transaction) => {
```
Comment (adrkakol) [inline src/comments/comments.controller.ts:71]: You have to check if !admin user has access to the place to which he creates a comment. Just use places service method isPlaceManageableByUser

File: src/comments/comments.controller.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      return comment;
+    });
+  }
+
+  @ApiResponse({ status: 204, description: 'deletes comment and returns empty response' })
+  @SetMetadata(MetadataKey.ALLOWED_ROLES, [UserRole.ADMIN, UserRole.PLACE_MANAGER])
+  @UseGuards(AuthGuard)
+  @HttpCode(HttpStatus.NO_CONTENT)
+  @Delete('/:id')
+  public async deleteComment(@Param('id') id: string): Promise<void> {
```
Comment (adrkakol) [inline src/comments/comments.controller.ts:87]: You have to check if !admin user has access to the place to which he creates a comment. Just use places service method isPlaceManageableByUser

File: src/comments/dto/create-comment.dto.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  @IsNotEmpty()
+  title!: string;
+
+  @ApiProperty()
+  @IsString()
+  @IsNotEmpty()
+  message!: string;
+
+  @ApiProperty()
+  @IsArray()
```
Comment (adrkakol) [inline src/comments/dto/create-comment.dto.ts:16]: Please check if swagger has managed to reflect the property of Array<String>. If not, add proper type to the `ApiPropertyDecorator`

File: src/comments/dto/update-comment.dto.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  @IsOptional()
+  @IsString()
+  @IsNotEmpty()
+  message?: string;
+
+  @ApiProperty()
+  @IsOptional()
+  @IsArray()
+  @IsNotEmpty()
+  links?: string;
```
Comment (adrkakol) [inline src/comments/dto/update-comment.dto.ts:21]: Same thing here, please make sure links type is reflected in swagger

File: migrations/20220324080020-add-place-comment.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
+'use strict';
+
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    await queryInterface.sequelize.transaction(async (transaction) => {
+      await queryInterface.createTable(
```
Comment (adrkakol) [inline migrations/20220324080020-add-place-comment.js:6]: I am pretty sure we also want to have updatedAt date in the database. Could you add it?

**Review by adrkakol — COMMENTED** [review]
Nice work with comments :) I have a few things for improvement

File: migrations/20220324080020-add-place-comment.js (hunk 2 above)
Comment (michaldomagalski) [reply migrations/20220324080020-add-place-comment.js:36]: Yup, that makes more sense.

File: migrations/20220324080020-add-place-comment.js (hunk 3 above)
Comment (michaldomagalski) [reply migrations/20220324080020-add-place-comment.js:37]: 👍

File: migrations/20220324080020-add-place-comment.js (hunk 1 above)
Comment (michaldomagalski) [reply migrations/20220324080020-add-place-comment.js:28]: 👍

File: migrations/20220324080020-add-place-comment.js (hunk 4 above)
Comment (michaldomagalski) [reply migrations/20220324080020-add-place-comment.js:31]: True 👍

File: migrations/20220324080020-add-place-comment.js (hunk 10 above)
Comment (michaldomagalski) [reply migrations/20220324080020-add-place-comment.js:6]: Migration ads both fields.

File: src/comments/dto/update-comment.dto.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import { ApiProperty } from '@nestjs/swagger';
+
+export class UpdateCommentDto {
+  @ApiProperty()
+  @IsOptional()
+  @IsString()
+  @IsNotEmpty()
+  title?: string;
+
+  @ApiProperty()
```
Comment (adrkakol) [inline src/comments/dto/update-comment.dto.ts:11]: Shouldn't optional property be set to null? Swagger shows it as required.
```suggestion
  @ApiProperty({ nullable: true })
```

Same about title and links

**Review by adrkakol — APPROVED** [review]
One last thing in the comment.

Thanks for load of great fixes - the code is more consistent now!

**Pull request merged**